iceoryx_posh/test/moduletests/test_capro_service.cpp

```diff
@@ -28,166 +28,649 @@ using namespace iox::capro;
+{
+    ServiceDescription serviceDescription1 = ServiceDescription();
+
+    EXPECT_EQ(serviceDescription1.getServiceID(), 0);
+    EXPECT_EQ(serviceDescription1.getEventID(), 0);
+    EXPECT_EQ(serviceDescription1.getInstanceID(), 0);
+    EXPECT_THAT(serviceDescription1.getServiceIDString(), StrEq("0"));
+    EXPECT_THAT(serviceDescription1.getEventIDString(), StrEq("0"));
+    EXPECT_THAT(serviceDescription1.getInstanceIDString(), StrEq("0"));
+}
+
+TEST_F(ServiceDescription_test, ServiceDescriptionStringCtorCreatesServiceDescriptionWithValuesPassedToTheCtor)
+{
+    IdString_t testService("1");
+    IdString_t testInstance("2");
+    IdString_t testEvent("3");
+    ServiceDescription::ClassHash testHash = {1, 2, 3, 4};
+
+    ServiceDescription serviceDescription1 = ServiceDescription(testService, testInstance, testEvent, testHash);
+
+    EXPECT_THAT(serviceDescription1.getServiceIDString(), StrEq("1"));
+    EXPECT_THAT(serviceDescription1.getInstanceIDString(), StrEq("2"));
+    EXPECT_THAT(serviceDescription1.getEventIDString(), StrEq("3"));
+    EXPECT_EQ(uint16_t(1), serviceDescription1.getServiceID());
+    EXPECT_EQ(uint16_t(2), serviceDescription1.getInstanceID());
+    EXPECT_EQ(uint16_t(3), serviceDescription1.getEventID());
+    EXPECT_EQ(uint32_t(1), serviceDescription1.getClassHash()[0]);
+    EXPECT_EQ(uint32_t(2), serviceDescription1.getClassHash()[1]);
+    EXPECT_EQ(uint32_t(3), serviceDescription1.getClassHash()[2]);
+    EXPECT_EQ(uint32_t(4), serviceDescription1.getClassHash()[3]);
+}
+TEST_F(ServiceDescription_test, ServiceDescriptionStringCtorWithNonIntegerStringValuesSetTheIDsToInvalid)
 {
     IdString_t testService("Service");
     IdString_t testInstance("Instance");
     IdString_t testEvent("Event");
     ServiceDescription::ClassHash testHash = {1, 2, 3, 4};
-    ServiceDescription csd = ServiceDescription(testService, testInstance, testEvent, testHash);
 
-    EXPECT_THAT(csd.getServiceIDString(), StrEq("Service"));
-    EXPECT_THAT(csd.getInstanceIDString(), StrEq("Instance"));
-    EXPECT_THAT(csd.getEventIDString(), StrEq("Event"));
-    EXPECT_EQ(uint32_t(1), csd.getClassHash()[0]);
-    EXPECT_EQ(uint32_t(2), csd.getClassHash()[1]);
-    EXPECT_EQ(uint32_t(3), csd.getClassHash()[2]);
-    EXPECT_EQ(uint32_t(4), csd.getClassHash()[3]);
+    ServiceDescription serviceDescription1 = ServiceDescription(testService, testInstance, testEvent, testHash);
+
+    EXPECT_EQ(InvalidID, serviceDescription1.getServiceID());
+    EXPECT_EQ(InvalidID, serviceDescription1.getInstanceID());
+    EXPECT_EQ(InvalidID, serviceDescription1.getEventID());
 }
 
-TEST_F(ServiceDescription_test, operatorEq)
+TEST_F(ServiceDescription_test, ServiceDescriptionStringCtorWithOutOfBoundaryIntegerStringValuesSetTheIDsToInvalid)
 {
-    EXPECT_TRUE(csd1 == csd1Eq);
-    EXPECT_TRUE(csd2 == csd2Eq);
-    EXPECT_TRUE(csd3 == csd3Eq);
-    EXPECT_TRUE(csd4 == csd4Eq);
+    IdString_t testService("65536");
```

Comment:
I didnt understand. testService is a string right ?

Reply:
This is correct, but magic numbers are not good to understand. My advise is to initialize the ServiceDescription here with uint instead of a string. This make more clear that you do an out-of-bounds check here.

Reply:
Yeah, true. But string is required to cover the branch of the constructor in lines 136, 142 and 148 of servicedescription.cpp. To cover those branch, I think the value has to be a string like "65536".

Reply:
Ah yes, i was misled by the two possible cases. Nevertheless there should be an explicit out-of-bound test that use the uint Ctor and `iox::capro::AnyService + 1` to verify that too.

Reply:
I'm getting overflow error while trying to build the test with iox::capro::AnyService + 1 as value to uint16_t

Reply:
Uh yes, the Ctor takes uint16_t as type. Then the test makes of course no sense. Please leave a comment in this test then where the value 65536 comes from. You can do that with `/// @note` as doxygen comment.